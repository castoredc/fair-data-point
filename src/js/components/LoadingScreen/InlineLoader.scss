@import '../../scss/colors.scss';

.Loader {
  padding: 50px 0;

  .LoadingSpinner {
    margin: auto;
  }

  &.Overlay {
    position: absolute;
    z-index: 999;
    left: 50%;
    margin-left: -32px;
  }
}