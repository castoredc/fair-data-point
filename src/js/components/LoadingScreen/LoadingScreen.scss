@import '../../scss/colors.scss';

.LoadingScreen {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: $bg;
  opacity: 0.95;
  z-index: 6;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  .LoadingSpinner {
    margin-bottom: 16px;
  }

  > h1 {
    font-size: 16px;
    font-weight: 600;
    line-height: 1.5;
    color: $n800;
  }
}
