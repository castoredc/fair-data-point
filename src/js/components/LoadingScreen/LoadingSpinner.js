import React from 'react';
import './LoadingSpinner.scss';

const LoadingSpinner = () => {
  return (
    <div className="LoadingSpinner">
      <div />
      <div />
      <div />
    </div>
  );
};

export default LoadingSpinner;
