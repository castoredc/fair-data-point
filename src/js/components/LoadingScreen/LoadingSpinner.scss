@import '../../scss/colors.scss';

.LoadingSpinner {
  width: 64px;
  height: 64px;

  > div {
    box-sizing: border-box;
    display: block;
    position: absolute;
    width: 51px;
    height: 51px;
    margin: 6px;
    border: 6px solid $blue400;
    border-radius: 50%;
    animation: spinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
    border-color: $blue400 transparent transparent transparent;

    &:nth-child(1) {
      animation-delay: -0.45s;
    }

    &:nth-child(2) {
      animation-delay: -0.3s;
    }

    &:nth-child(3) {
      animation-delay: -0.15s;
    }
  }
}

@keyframes spinner {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
