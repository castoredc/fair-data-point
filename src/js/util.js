export const by = key => (object, item) => {
  object[item[key]] = item;
  return object;
};

export const classNames = (...args) => args.filter(identity).join(' ');

export const debounce = (fn, delay) => {
  let timer;
  return (...args) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      fn(...args);
    }, delay);
  };
};

export const identity = x => x;

export const log = (value, prefix = '') => {
  console.log(prefix, value);
  return value;
};

export const isNumeric = number =>
  !isNaN(parseFloat(number)) && isFinite(number);

export const preventDefault = e => {
  e.preventDefault();
};

export const localizedText = (texts, language = 'en') => {
  for (const text of texts) {
    if(text.language === language) return text.text;
  }
  return texts[0].text;
};
