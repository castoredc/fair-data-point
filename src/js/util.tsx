import React, { cloneElement } from 'react';

export const by = key => (object, item) => {
    object[item[key]] = item;
    return object;
};

export const classNames = (...args) => args.filter(identity).join(' ');

export const debounce = (fn, delay) => {
    let timer;
    return (...args) => {
        clearTimeout(timer);
        timer = setTimeout(() => {
            fn(...args);
        }, delay);
    };
};

export const identity = x => x;

export const log = (value, prefix = '') => {
    console.log(prefix, value);
    return value;
};

export const isNumeric = number => !isNaN(parseFloat(number)) && isFinite(number);

export const preventDefault = e => {
    e.preventDefault();
};

export const paragraphText = text => {
    return text.split('\n').map((item, i) => {
        return item.length > 1 ? <p key={i}>{item}</p> : null;
    });
};

export const localizedText = (texts, language = 'en', paragraph = false) => {
    if (typeof texts === 'string') {
        return texts;
    }

    if (texts === null || (Array.isArray(texts) && texts.length === 0) || typeof texts === 'undefined') {
        return '';
    }

    for (const text of texts) {
        if (text.language === language) {
            if (paragraph) {
                return paragraphText(text.text);
            }
            return text.text;
        }
    }

    if (paragraph) {
        return paragraphText(texts[0].text);
    }

    return texts[0].text;
};

export const isURL = str => {
    if (typeof str !== 'string') {
        return false;
    }

    var pattern = new RegExp(
        '^(https?:\\/\\/)?' + // protocol
            '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
            '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
            '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
            '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
            '(\\#[-a-z\\d_]*)?$',
        'i'
    ); // fragment locator
    return !!pattern.test(str);
};

export const getCenterFromDegrees = data => {
    if (!(data.length > 0)) {
        return false;
    }

    let num_coords = data.length;

    let X = 0.0;
    let Y = 0.0;
    let Z = 0.0;

    for (let i = 0; i < data.length; i++) {
        let lat = (data[i][0] * Math.PI) / 180;
        let lon = (data[i][1] * Math.PI) / 180;

        let a = Math.cos(lat) * Math.cos(lon);
        let b = Math.cos(lat) * Math.sin(lon);
        let c = Math.sin(lat);

        X += a;
        Y += b;
        Z += c;
    }

    X /= num_coords;
    Y /= num_coords;
    Z /= num_coords;

    let centerLon = Math.atan2(Y, X);
    let centerHyp = Math.sqrt(X * X + Y * Y);
    let centerLat = Math.atan2(Z, centerHyp);

    let newX = (centerLat * 180) / Math.PI;
    let newY = (centerLon * 180) / Math.PI;

    return [newX, newY];
};

export const replaceAt = (array, index, value) => {
    const ret = array.slice(0);
    ret[index] = value;
    return ret;
};

export const mergeData = (defaultData, newData) => {
    return Object.keys(defaultData).reduce((a, key) => ({ ...a, [key]: key in newData ? newData[key] : defaultData[key] }), defaultData);
};

export const ucfirst = text => {
    return text.charAt(0).toUpperCase() + text.slice(1);
};

export const downloadFile = (contents, filename) => {
    const nav = window.navigator as any;

    if (!nav.msSaveOrOpenBlob) {
        const url = window.URL.createObjectURL(new Blob([contents]));
        const link = document.createElement('a');
        link.href = url;
        link.setAttribute('download', filename);
        document.body.appendChild(link);
        link.click();
    } else {
        // IE
        const url = nav.msSaveOrOpenBlob(new Blob([contents]), filename);
    }
};

export const cloneIfComposite = (children, props) => {
    return children && children.type && typeof children.type !== 'string' ? cloneElement(children, props) : children;
};

export const getType = type => {
    switch (type) {
        case 'data-model':
            return 'data model';
        case 'metadata-model':
            return 'metadata model';
        default:
            return '';
    }
};

export const findOptionByValue = (value, options) => {
    for (const option of options) {
        // Check if it's a grouped option
        if (option.options) {
            // Search within this group
            const foundOption = option.options.find(opt => opt.value === value);
            if (foundOption) return foundOption;
        } else {
            // It's a non-grouped option
            if (option.value === value) return option;
        }
    }
    return null; // Return null if not found
};
