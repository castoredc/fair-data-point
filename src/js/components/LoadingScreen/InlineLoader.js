import React from 'react';
import './InlineLoader.scss';
import LoadingSpinner from "./LoadingSpinner";
import {classNames} from "../../util";

export default ({overlay = false}) => {
    return (
        <div className={classNames('Loader', overlay && 'Overlay')}>
            <LoadingSpinner />
        </div>
    );
};
