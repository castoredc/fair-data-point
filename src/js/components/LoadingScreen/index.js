import React, { Component } from 'react';
import LoadingSpinner from './LoadingSpinner';
import EventListener from '../EventListener';
import { preventDefault } from '../../util';
import './LoadingScreen.scss';

const MINIMUM_DURATION = 2500;
const LENGTHY_OPERATION = 6000;

const INITIAL_STATE = {
  canClose: false,
  lengthyOperation: false,
  open: false,
};

class LoadingScreen extends Component {
  state = {
    ...INITIAL_STATE,
    open: this.props.showLoading,
  };

  componentDidMount() {
    if (this.state.open) {
      this.showLoadingScreen();
    }
  }

  showLoadingScreen = () => {
    this.startTimers();
    this.setState({ open: true });
  };

  closeLoadingScreenIfPossible = () => {
    if (this.state.canClose) {
      this.setState(INITIAL_STATE);
      this.clearTimers();
    }
  };

  startTimers = () => {
    this.minimumDurationTimeout = setTimeout(() => {
      this.setState({ canClose: true });
      if (!this.props.showLoading) {
        this.closeLoadingScreenIfPossible();
      }
    }, MINIMUM_DURATION);

    this.lengthyOperationTimeout = setTimeout(() => {
      this.setState({ lengthyOperation: true });
    }, LENGTHY_OPERATION);
  };

  componentDidUpdate(prevProps) {
    const { showLoading } = this.props;
    const { open } = this.state;

    if (prevProps.showLoading === showLoading) {
      return;
    }

    if (!open && showLoading) {
      this.showLoadingScreen();
      return;
    }

    if (open && !showLoading) {
      this.closeLoadingScreenIfPossible();
      return;
    }
  }

  clearTimers = () => {
    clearTimeout(this.minimumDurationTimeout);
    clearTimeout(this.lengthyOperationTimeout);
  };

  componentWillUnmount() {
    this.clearTimers();
  }

  render() {
    const { open, lengthyOperation } = this.state;
    const { message } = this.props;

    if (!open) {
      return null;
    }

    return (
      <section className="LoadingScreen" aria-live="assertive" role="alert">
        <EventListener
          target={document}
          type="keydown"
          listener={preventDefault}
        />
        <LoadingSpinner />
        <h1>
          {lengthyOperation
            ? 'Oops, this is slower than expected. Please wait a bit longer.'
            : message}
        </h1>
      </section>
    );
  }
}

export default LoadingScreen;
