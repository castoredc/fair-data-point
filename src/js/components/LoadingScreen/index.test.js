import React from 'react';
import { shallow, mount } from 'enzyme';
import LoadingScreen from '../LoadingScreen';
import renderer from 'react-test-renderer';
import { expectSnapshotMatch } from '../../../test/util';

const INITIAL_MESSAGE = 'Your action is pending...';

beforeEach(() => {
  jest.useFakeTimers();
});

afterEach(() => {
  jest.clearAllTimers();
});

describe('<LoadingScreen />', () => {
  it('Renders null when initially passing showLoading prop', () => {
    const loadingScreen = <LoadingScreen showLoading={false} />;
    expectSnapshotMatch(loadingScreen);
  });

  it('Renders loading screen when passing showLoading prop', () => {
    const loadingScreen = (
      <LoadingScreen showLoading={true} message={INITIAL_MESSAGE} />
    );
    expectSnapshotMatch(loadingScreen);
  });

  it('Is displayed for a minimum 2.5 seconds initially open', () => {
    const loadingScreen = shallow(<LoadingScreen showLoading={true} />);
    loadingScreen.setProps({ showLoading: false });

    // Loading screen is still shown after showLoading prop set to false
    expect(loadingScreen.getElement()).not.toBeNull();

    // 2 seconds not enough to close the screen
    jest.advanceTimersByTime(2000);
    expect(loadingScreen.getElement()).not.toBeNull();

    jest.advanceTimersByTime(501);

    // Loading screen is now closed
    expect(loadingScreen.getElement()).toBeNull();
  });

  it('Is displayed for a minimum 2.5 seconds initially closed', () => {
    const loadingScreen = shallow(<LoadingScreen />);
    loadingScreen.setProps({ showLoading: true });
    loadingScreen.setProps({ showLoading: false });

    // Loading screen is still shown after showLoading prop set to false
    expect(loadingScreen.getElement()).not.toBeNull();

    // 2 seconds not enough to close the screen
    jest.advanceTimersByTime(2000);
    expect(loadingScreen.getElement()).not.toBeNull();

    jest.advanceTimersByTime(501);

    // Loading screen is now closed
    expect(loadingScreen.getElement()).toBeNull();
  });

  it('Is closeable instantly at any time after the initial 2.5 seconds initially open', () => {
    const loadingScreen = shallow(<LoadingScreen showLoading={true} />);

    jest.advanceTimersByTime(3000);
    loadingScreen.setProps({ showLoading: false });

    expect(loadingScreen.getElement()).toBeNull();
  });

  it('Is closeable instantly at any time after the initial 2.5 seconds initially closed', () => {
    const loadingScreen = shallow(<LoadingScreen showLoading={false} />);
    loadingScreen.setProps({ showLoading: true });
    loadingScreen.setProps({ showLoading: false });

    jest.advanceTimersByTime(3000);
    loadingScreen.setProps({ showLoading: false });

    expect(loadingScreen.getElement()).toBeNull();
  });

  it('Can be closed and and re-opened, initially open', () => {
    const loadingScreen = shallow(<LoadingScreen showLoading={true} />);
    expect(loadingScreen.getElement()).not.toBeNull();

    jest.advanceTimersByTime(3000);
    loadingScreen.setProps({ showLoading: false });

    // Closed
    expect(loadingScreen.getElement()).toBeNull();

    loadingScreen.setProps({ showLoading: true });

    // Re-opened
    expect(loadingScreen.getElement()).not.toBeNull();
  });

  it('Can be opened, closed and and re-opened, initially closed', () => {
    const loadingScreen = shallow(<LoadingScreen showLoading={false} />);
    expect(loadingScreen.getElement()).toBeNull();

    loadingScreen.setProps({ showLoading: true });

    //Opened
    expect(loadingScreen.getElement()).not.toBeNull();

    jest.advanceTimersByTime(3000);
    loadingScreen.setProps({ showLoading: false });

    // Closed
    expect(loadingScreen.getElement()).toBeNull();

    loadingScreen.setProps({ showLoading: true });

    // Re-opened
    expect(loadingScreen.getElement()).not.toBeNull();
  });

  it('Displays originally passed message for first 6 seconds', () => {
    const loadingScreen = renderer.create(
      <LoadingScreen showLoading={true} message={INITIAL_MESSAGE} />
    );

    expect(loadingScreen).toMatchSnapshot();
  });

  it('Displays lengthy operation message after the loading screen has been active for 6 seconds', () => {
    const loadingScreen = renderer.create(
      <LoadingScreen showLoading={true} message={INITIAL_MESSAGE} />
    );

    jest.advanceTimersByTime(6001);
    expect(loadingScreen).toMatchSnapshot();
  });
});
