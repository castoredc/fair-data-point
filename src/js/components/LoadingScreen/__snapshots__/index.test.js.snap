// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadingScreen /> Displays lengthy operation message after the loading screen has been active for 6 seconds 1`] = `
<section
  aria-live="assertive"
  className="LoadingScreen"
  role="alert"
>
  <div
    className="LoadingSpinner"
  >
    <div />
    <div />
    <div />
  </div>
  <h1>
    Oops, this is slower than expected. Please wait a bit longer.
  </h1>
</section>
`;

exports[`<LoadingScreen /> Displays originally passed message for first 6 seconds 1`] = `
<section
  aria-live="assertive"
  className="LoadingScreen"
  role="alert"
>
  <div
    className="LoadingSpinner"
  >
    <div />
    <div />
    <div />
  </div>
  <h1>
    Your action is pending...
  </h1>
</section>
`;

exports[`<LoadingScreen /> Renders loading screen when passing showLoading prop 1`] = `
<section
  aria-live="assertive"
  className="LoadingScreen"
  role="alert"
>
  <div
    className="LoadingSpinner"
  >
    <div />
    <div />
    <div />
  </div>
  <h1>
    Your action is pending...
  </h1>
</section>
`;

exports[`<LoadingScreen /> Renders null when initially passing showLoading prop 1`] = `null`;
